File: hw1/p12.py
# e7ee42d4-1251-4d61-a61e-c67ea784c217


def divideConsecutive(nums, k):
    if len(nums) == 0:
        return "true"
    
    if len(nums) % k != 0:
        return "false"
    
    minNum = min(nums)
    
    nums.sort()
    
    if k % 2 == 0:
        for i in range(k // 2):
            if minNum + i in nums:
                nums.remove(minNum + i)
            else:
                return "false"
             
            if minNum + k - 1 - i in nums:
                nums.remove(minNum + k - 1 - i)
            else:
                return "false"
    else:
        for i in range(k // 2):
            if minNum + i + 1 == minNum + k - 2 - i and minNum+i+1 in nums:
                nums.remove(minNum + i + 1)

            if minNum + i in nums:
                nums.remove(minNum + i)
            else:
                return "false"
            
            if minNum + k - 1 - i in nums :
                nums.remove(minNum + k - 1 - i)
            else:
                return "false"

    
    return divideConsecutive(nums, k)

def main():
    array = list(map(int, input().split()))
    k = int(input());

    print(divideConsecutive(array,k))
    
main()


'''
1 2 3 4 5 6 7 8 9
3

for 0 in range(2)



'''

File: hw1/p12.cpp
/*
Given an array of positive integers nums and a positive integer k, check whether it is 
possible to divide nums into sets of k consecutive numbers.

Write a recursive function to solve this problem. Return true if it is possible. 
Otherwise, return false. The input has 2 lines. The first line is the array of positive 
integers nums. The second line is the positive integer k.
*/

// e3fc3919-8de9-4d8d-a274-496b820bc2cf

#include <iostream>
#include <vector>
#include <algorithm>
#include <sstream>

using namespace std;

bool divideConsecutive(vector<int>& nums, int k) {
    if (nums.empty()) {
        return true;
    }

    if (nums.size() % k != 0) {
        return false;
    }

    int minNum = *min_element(nums.begin(), nums.end());

    for (int i = 0; i < k; ++i) {
        
        auto it = find(nums.begin(), nums.end(), minNum + i);

        if (it != nums.end()) {
            nums.erase(it); 
        } else {
            return false;
        }
    }

    return divideConsecutive(nums, k);
}

int main() {
    
    vector<int> nums;
    int k;
    string temp;
    int intTemp;

    getline(cin,temp);

    stringstream ss(temp);

    while(ss >> intTemp){
        nums.push_back(intTemp);
    }
    cin >> k;
    

    if (divideConsecutive(nums, k)) {
        cout << "true" << endl;
    } else {
        cout << "false" << endl;
    }

    return 0;
}

